### PR TITLE
fix: Morphtarget Export and Import

### PR DESCRIPTION
**Pull request by effs**
## Highlights

Export: produces correct data so that it actually *can* be imported correctly\*
Import: produces working morphtargets for vanilla/modified/custom morphs\*\*

## Breaking

Previously exported and/or imported morphtargets are likely to have at least some issues. Users may need to re-export and port their work on top of that.

## More

The commits are kinda split up and the commit messages worth reading but, YMMV, it may be easier to actually review the code by reading the entire file instead of a diff.

\*,\*\* I mean, like, _probably_? I haven't tested that extensively.

_Evidence for the defense: sculpted neck on a head morph (with still-working customization & expressions)_
image

### CODE REVIEW

**Comment by effs** [comment]
I still need to tidy some things up a bit but I think the logic is there and uncluttered enough to review.

File: WolvenKit.Modkit/RED4/Tools/MorphTargetTools.cs [hunk 1]
```diff
@@ -199,35 +199,44 @@ private static TargetsInfo GetTargetInfos(CR2WFile cr2w, int subMeshC)
...
+                var storedMapsCount = numVertexDiffsMappingInEachChunk[subMeshIndex];
 
-                if (i == 0)
-                {
-                    diffsbuffer.Position = targetStartsInVertexDiffs * 12;
-                }
-                else
-                {
-                    diffsbuffer.Position = targetStartsInVertexDiffs * 12;
+                // I *think* this will only ever be >, but just in case
```
Comment (seberoth) [inline WolvenKit.Modkit/RED4/Tools/MorphTargetTools.cs:217]: Not needed but `diffsCount / 2 - storedMapsCount`. Maybe could add a check then if its > 1 or < -1?

**Review by seberoth — COMMENTED** [review]
rest looks goot to me

File: WolvenKit.Modkit/RED4/Tools/MorphTargetImportTools.cs [hunk 2]
```diff
@@ -106,256 +88,320 @@ public bool ImportMorphTargets(FileInfo inGltfFile, Stream intargetStream, GltfI
...
+                        var tangentAs10BitShiftedInt = Converters.Vec4ToU32(zUpTangentDelta);
 
-                        // Tangent
-                        output = Converters.Vec4ToU32(new Vec4(tangentsDelta.X, -tangentsDelta.Z, tangentsDelta.Y, 1f));
-                        diffsWriter.Write(output);
+                        // 4 + 4 + 4 bytes per diff, no padding
+                        diffsWriter.Write(positionAs10BitUnsignedInt);
+                        diffsWriter.Write(normalAs10BitShiftedInt);
+                        diffsWriter.Write(tangentAs10BitShiftedInt);
                     }
```
[the remark this answers is not in the document]
Comment (effs) [reply WolvenKit.Modkit/RED4/Tools/MorphTargetImportTools.cs:344]: True - I think I prefer the current, it's mappings' responsibility to handle its own alignment.

**Review by Syrrr — APPROVED** [review]
I approve this message.

**Review by seberoth — APPROVED** [review]
lgtm

**Pull request merged**